controller-server/src/main/java/com/yahoo/vespa/hosted/controller/restapi/deployment/DeploymentApiHandler.java

```diff
@@ -129,25 +124,73 @@ private HttpResponse root(HttpRequest request) {
+            for (Run run : statistics.runningUpgrade()) {
                 Cursor applicationObject = runningArray.addObject();
                 toSlime(applicationObject, run.id().application(), request);
                 applicationObject.setString("running", run.id().type().jobName());
             }
+
+            class RunInfo { //  ヽ༼ຈل͜ຈ༽━☆ﾟ.*･｡ﾟ
+                final Run run;
+                final boolean upgrade;
+                RunInfo(Run run, boolean upgrade) { this.run = run; this.upgrade = upgrade; }
+                @Override public String toString() { return run.id().toString(); }
+            }
+            Cursor instancesArray = versionObject.setArray("applications");
+            Stream.of(statistics.failingUpgrades().stream().map(run -> new RunInfo(run, true)),
+                      statistics.otherFailing().stream().map(run -> new RunInfo(run, false)),
+                      statistics.runningUpgrade().stream().map(run -> new RunInfo(run, true)),
+                      statistics.otherRunning().stream().map(run -> new RunInfo(run, false)),
+                      statistics.productionSuccesses().stream().map(run -> new RunInfo(run, true)))
+                  .flatMap(identity())
+                  .collect(Collectors.groupingBy(run -> run.run.id().application(),
+                                                 LinkedHashMap::new, // Put apps with failing and running jobs first.
+                                                 groupingBy(run -> run.run.id().type())))
+                  .forEach((instance, runs) -> {
+                      Cursor instanceObject = instancesArray.addObject();
+                      instanceObject.setString("tenant", instance.tenant().value());
+                      instanceObject.setString("application", instance.application().value());
+                      instanceObject.setString("instance", instance.instance().value());
+                      instanceObject.setLong("productionJobCount", jobsByInstance.get(instance).production().size());
+                      instanceObject.setString("upgradePolicy", toString(deploymentStatuses.matching(status -> status.application().id().equals(TenantAndApplicationId.from(instance)))
+                                                                                           .first().map(status -> status.application().deploymentSpec())
+                                                                                           .flatMap(spec -> spec.instance(instance.instance()).map(DeploymentInstanceSpec::upgradePolicy))
+                                                                                           .orElse(DeploymentSpec.UpgradePolicy.defaultPolicy)));
+                      Cursor allJobsObject = instanceObject.setObject("allJobs");
+                      Cursor upgradeJobsObject = instanceObject.setObject("upgradeJobs");
+                      for (JobType type : JobType.allIn(controller.system())) {
```

Comment:
Would be nice if we could just list relevant jobs, and maybe list all the jobs once in a list for order, if we need that.

Reply:
I guess that's an option that saves some JSON over the wire :)